fontforgeexe/charview.c

```diff
@@ -6638,22 +6664,32 @@ static void CVMenuClose(GWindow gw, struct gmenuitem *UNUSED(mi), GEvent *UNUSED
 	GDrawDestroyWindow(gw);
 }
 
+// This can be triggered two ways: by the "X" buttons on tabs, and regularly in the menu.
 static void CVMenuCloseTab(GWindow gw, struct gmenuitem *UNUSED(mi), GEvent *UNUSED(e)) {
     CharView *cv = (CharView *) GDrawGetUserData(gw);
     int pos, i;
 
     if ( cv->b.container || cv->tabs==NULL || cv->former_cnt<=1 )
 return;
-    pos = GTabSetGetSel(cv->tabs);
+    pos = cv->ctpos == -1 ? GTabSetGetSel(cv->tabs) : cv->ctpos;
     free(cv->former_names[pos]);
-    for ( i=pos+1; i<cv->former_cnt; ++i ) {
-	cv->former_names[i-1] = cv->former_names[i];
-    cv->cvtabs[i-1] = cv->cvtabs[i];
+    for ( i=pos; i<cv->former_cnt; ++i ) {
```

Comment:
To test more easily, I set `CV_TABMAX` to 5, and there's no invalid write, even when `cv->former_cnt` is 5, `i` will never get to be more than 4.

Reply:
If there are five entries then those are indexed 0 through 4, and (the new) line 6678 is

```
cv->cvtabs[i] = cv->cvtabs[i+1];
```
It's the change to `i+1` without a change to the max `i` that I'm worried about.

Reply:
Note that I could be missing some reason this is OK after all. But it's also possible that different memory allocation systems and alignments will make this usually OK but occasionally segfault-y. I can't yet see "by analysis" why it's OK to access one further index than we were before.

Reply:
`cvtabs` is a fixed-size array of size `charview_cvtabssz+1` (101) so it won't be an invalid access for that, so only `former_names` is in question here

Reply:
I'm sorry, a lot of this is spaghetti. I renamed this constant in the first place because `FORMER_MAX` was such a confusing name.

I think **either**:
* a `+1` should be added to `views.h` line 302 to make them match.
* `CV_TABMAX` a/k/a `FORMER_MAX` should be done away with altogether and replaced with `charview_cvtabssz`

Reply:
Just so I understand, what would be the problem with changing the loop to
`for ( i=pos+1; i<cv->former_cnt-1; ++i ) {` ? Given the use of `i+1` why wouldn't that have the intended result?

Reply:
My version sets i=pos

Reply:
Sorry, yes, I copied the wrong line. It was `for ( i=pos+1; i<cv->former_cnt; ++i ) {`, you've changed it to `for ( i=pos; i<cv->former_cnt; ++i ) {`. Because yours starts at one less than the old loop but ends at the same place, you're iterating over one more value. That's the heart of what I'm asking about in the initial comment: Is that extra `i` value accidental or needed, and if the latter what is the explanation?

Reply:
You're correct, the change is unnecessary. I have no idea why this doesn't cause a crash as I'd expect it to. I have a feeling it's partly because `cvtabs` won't always be equal in size to `former_names`. 

I'll revert it.

Reply:
It's very unlikely to cause a crash because it's very unlikely that the area immediately after an array with small entries will be unmapped. Usually you'll just wind up copying garbage into the top slot, which is harmless because it isn't considered part of the "count" anyway. 

The primary problem with this sort of thing is readability. Someone may come by wondering what the extra copy is for and worry they're missing something.